pkg/limit/limiter_test.go

```diff
@@ -376,11 +390,35 @@ func TestBatchExpiration(t *testing.T) {
 		}
 
 		assert.Equal(t, totalKeys, o.len())
+
+		// Wait for immediate callbacks to be processed
+		for i := 0; i < totalKeys; i++ {
+			select {
+			case <-completionChan:
+			case <-time.After(100 * time.Millisecond):
+				t.Fatalf("Timeout waiting for immediate callback %d", i)
+			}
+		}
+
 		time.Sleep(expireInterval / 2)
+
+		// Wait for each batch with both count checking and completion tracking
 		for i := range batchNum {
-			assert.Equal(t, totalKeys+expireBatchSize*i, o.len())
-			time.Sleep(expireInterval)
+			expectedCount := totalKeys + expireBatchSize*(i+1)
+
+			// First, wait for the callbacks to actually execute
+			for j := 0; j < expireBatchSize; j++ {
```

Comment:
I think it would be good to use 'range' here as well.
```suggestion
			for j := range expireBatchSize {
```